### PR TITLE
GLM PyTorch: add non-distributed training support, fix fp32 training and remove apex dependency from benchmarks directory

### PR DESCRIPTION
**Pull request by dynamicheart**
…and remove apex dependency from benchmarks directory

### CODE REVIEW

File: training/benchmarks/glm/pytorch/schedulers/base.py [hunk 1]
```diff
@@ -54,8 +54,8 @@ def get_lr(self):
             return float(self.start_lr) * self.num_iters / self.warmup_iter
         else:
             if self.decay_style == self.DECAY_STYLES[0]:
-                decay_step_ratio = (self.num_iters -
-                                    self.warmup_iter) / self.end_iter
+                decay_step_ratio = min(
```
Comment (dynamicheart) [inline training/benchmarks/glm/pytorch/schedulers/base.py:57]: prevent learning_rate from becoming negative

File: training/benchmarks/glm/pytorch/dataloaders/dataloader.py [hunk 2]
```diff
@@ -168,9 +168,10 @@ def build_data_loader(dataset,
     """Data loader. Note that batch-size is the local (per GPU) batch-size."""
     if worker_init_fn is None:
         worker_init_fn = WorkerInitializer.default()
-    world_size = dist.get_world_size()
+    rank = dist_pytorch.get_rank()
+    world_size = dist_pytorch.get_world_size()
     sampler = torch.utils.data.distributed.DistributedSampler(
-        dataset, num_replicas=world_size, shuffle=shuffle)
+        dataset, rank=rank, num_replicas=world_size, shuffle=shuffle)
```
Comment (dynamicheart) [inline training/benchmarks/glm/pytorch/dataloaders/dataloader.py:174]: Passing in `rank` to `DistributedSampler` , ensuring that `DistributedSampler` will not invoke `torch.distributed.get_rank()` and can still function properly in single device training scenarios where torch distributed group is not initialized.

File: training/benchmarks/glm/pytorch/train/trainer_adapter.py [hunk 3]
```diff
@@ -6,7 +6,10 @@
 from typing import Tuple
 
 import optimizers
-from apex.optimizers import FusedAdam as Adam
+try:
+    from apex.optimizers import FusedAdam as Adam
```
Comment (dynamicheart) [inline training/benchmarks/glm/pytorch/train/trainer_adapter.py:10]: The docker environment for other vendors may not have the apex library

File: training/benchmarks/glm/pytorch/train/trainer.py [hunk 4]
```diff
@@ -136,8 +136,9 @@ def train_one_epoch(self, dataloader):
                     self.config.gradient_accumulation_steps) / step_total_time
                 other_state["seq/s"] = sequences_per_second
 
-            loss_scale = self.optimizer.loss_scaler.loss_scale
-            other_state['loss_scale'] = loss_scale
+            if hasattr(self.optimizer, 'loss_scaler'):
+                loss_scale = self.optimizer.loss_scaler.loss_scale
+                other_state['loss_scale'] = loss_scale
```
Comment (dynamicheart) [inline training/benchmarks/glm/pytorch/train/trainer.py:142]: the optimizer doesn't have a loss scaler when running a fp32 training

File: training/benchmarks/glm/pytorch/schedulers/base.py (hunk 1 above)
Comment (yuzhou03) [reply training/benchmarks/glm/pytorch/schedulers/base.py:57]: it works

**Comment by yuzhou03** [comment]
if you make changes for standard case, please make sure:
1. changes are necessary. such as  add features, fix bugs,  improve performance,  readability etc.
2. attach running logs or provide a summary table of  running processes  that can prove the running process of  existed **nvidia** case config can be converged to target_accurracy are OK.
3. make sure all the vendor's adaption of standard case can still work for your changes, both attach running logs on these vendor's machine or provide a summary table of  running processes are OK.

File: training/benchmarks/glm/pytorch/train/trainer.py (hunk 4 above)
Comment (yuzhou03) [reply training/benchmarks/glm/pytorch/train/trainer.py:142]: LGTM

File: training/benchmarks/glm/pytorch/train/trainer_adapter.py (hunk 3 above)
Comment (yuzhou03) [reply training/benchmarks/glm/pytorch/train/trainer_adapter.py:10]: @upvenly plz double-check if this changes workable for other vendor

**Comment by dynamicheart** [comment]
[GLM_TORCH_DEMO_A100_1X8.log](https://github.com/FlagOpen/FlagPerf/files/10918044/GLM_TORCH_DEMO_A100_1X8.log)
[GLM_TORCH_DEMO_A100_1X2.log](https://github.com/FlagOpen/FlagPerf/files/10918045/GLM_TORCH_DEMO_A100_1X2.log)
